Add security model document (#8369)
hashicorp/nomad: website/pages/docs/internals/security.mdx

    to enforce further restrictions on operators. This is used to augment the
    built-in ACL system for fine-grained control over jobs.

* **[Namespaces](/guides/governance-and-policy/namespaces.html)**
    (**Enterprise Only**) - This feature allows for a cluster to be shared by
    multiple teams within a company. Using this logical separation is important
    for multi-tenant clusters to prevent users without access to that namespace
    from conflicting with each other. This requires ACLs to be enabled in order
    to be enforced.

* **[Resource Quotas](/guides/governance-and-policy/quotas.html)**
    (**Enterprise Only**) - Can limit a namespace’s access to the underlying
    compute resources in the cluster by setting upper-limits for operators.
    Access to these resource quotas can be managed via ACLs to ensure read-only
    access for operators so they can’t just change their quotas.

#### Recommendations

The following are security recommendations that can help significantly improve
the security of your cluster depending on your use case. We recommend always
practicing defense in depth when architecting the security mechanisms for your
environment.

> * **[Rotate Credentials](/docs/job-specification/vault.html)** -

Author: I'll massage this a bit further.  I think the intention is to recommend rotating credentials as much as possible, with these two examples being prominent examples.

    create and manage dynamic, rotated credentials is highly recommended to
    prevent secrets from being easily exposed within the [job
    specification](/docs/job-specification/index.html)
    itself which may be leaked into version control or otherwise be accidently
    stored on disk on an operator’s local machine. It is also possible to
    [integrate with Vault’s PKI secret engine](/guides/security/vault-pki-integration.html)
    to automatically generate and renew dynamic, unique X.509 certificates for
    each Nomad node with a short
    [TTL](https://en.wikipedia.org/wiki/Time_to_live).

* **[Running without Root](https://groups.google.com/forum/#!topic/nomad-tool/pSyMwC_FSFA)** -
    Certain features of Nomad can be used without needing to run the Nomad agent
    server or client as the `root` user. Instead you can granularly assign the
    appropriate capabilities in various ways for your Nomad agents. For example:
    Nomad servers only require access to the data directory; it is possible to
    use Nomad to orchestrate Docker containers by adding a non-root `nomad` user
    to the `docker` group to access the [default unix

>     socket](https://docs.docker.com/engine/reference/commandline/dockerd/#daemon-socket-option).

Reviewer: Making this a "recommendation" makes me nervous as there are tradeoffs involved. As specified in the linked message, running as unprivileged user prevents Nomad from using an in-memory filesystem for the per-task `/secrets` directory which causes all secrets to be persisted to disk.

Likewise running `raw_exec` with an unprivileged Nomad agent is actually _less_ safe than running `raw_exec` with a root Nomad agent. Since `raw_exec` tasks can only run as the same user as Nomad when Nomad is not root, they have full access to Nomad's data and configuration directories which may contain secrets that would not be divulged otherwise.

That being said: running `raw_exec` is *never* secure, so perhaps that example isn't valid.

I think running Nomad as non-root is only useful for functionally reducing jobs' privilege level to Docker's and therefore easing threat modelling and hardening in Docker-only environments.

Podman's rootless containers may make this a stronger recommendation since they provide task isolation while avoiding executing any privileged code (unlike dockerd). Thus simplifying the threat model by always providing tasks a _subset_ of Nomad's capabilities.

Again I'm fine with publishing this as-is, since I think it's even argue to pros of a non-root Nomad outweigh the cons.

Author: I'll restrict this recommendation to servers. It's definitely hard to run clients without being root.

Reviewer: Oh that's perfect. There's no reason for servers to run as root, and I don't think we make that clear anywhere.

* **Containers with Sandbox Runtimes** - In some situations, such as running
    untrusted code as a service, it may be worth considering using different
    container runtimes such as [gVisor](https://gvisor.dev/) or [Kata
    Containers](https://katacontainers.io/). These types of runtimes provide
    sandboxing features which help prevent raw access to the underlying shared

>     kernel for other containers and the Nomad client agent itself.

Reviewer: We should include a link to how to do that: https://www.nomadproject.io/docs/drivers/docker#runtime

* **[Disable Unused Drivers](/docs/configuration/client#driver-blacklist)** -
    Each driver provides different degrees of isolation, and bugs may allow
    unintended privilege escalation. If a task driver is not needed, you can
    disable it to reduce risk.

* **Linux Security Modules** - Use of security modules that can be directly
    integrated into operating systems such as AppArmor, SElinux, and Seccomp on
    both the Nomad hosts and applied to containers for an extra layer of
    security. Seccomp profiles are able to be passed directly to containers
    using the
    **[security_opt](/docs/drivers/docker.html#security_opt)**
    parameter available in the default [Docker
    driver](/docs/drivers/docker.html).

* **[Service Mesh](https://www.hashicorp.com/resources/service-mesh-microservices-networking)** -
    Integrating service mesh technologies such as
    **[Consul](https://www.consul.io/)** can be extremely useful for limiting
    and efficiently load balancing network connectivity within a cluster.

### Threat Model

The following are parts of the Nomad threat model:

> * **Nomad agent-to-agent communication** - Transport encryption for

Reviewer: I think `User-to-agent communication (HTTP)` would make a good followup to this and describe how that's protected by mTLS and ACL tokens.
```suggestion
* **Nomad agent-to-agent communication (RPC)** - Transport encryption for
```

Not sure about specifying "HTTP" vs "RPC" ... might be a needlessly confusing implementation detail.

Author: umm.. what would go under "User-to-agent communication" is currently covered under the subsequent sections "Tampering of data in transit" and "Access to data without authentication or authorization".  Any suggestions for how to restructure them?

    UDP based protocols within Nomad provide different mechanisms for enabling
    encryption including symmetric (shared gossip encryption keys) and
    asymmetric keys (TLS).

* **Tampering of data in transit** - Any tampering should be detectable via mTLS
    and cause Nomad to avoid processing the request.

* **Access to data without authentication or authorization** - Requests to the
    server should be authenticated and authorized using mTLS and ACLs
    respectively.

* **State modification or corruption due to malicious messages** - Improperly
    formatted messages are discarded while properly formatted messages require
    authentication and authorization.

* **Non-server members accessing raw data** - All servers that join the cluster
    require proper authentication and authorization in order to begin
    participating in Raft. All data in Raft should be encrypted with TLS.

* **Denial of Service against a node** - DoS attacks against a single node
    should not compromise the security posture of Nomad.

The following are not part of the threat model for server agents:

* **Access (read or write) to the Nomad data directory** - Information about the

>     jobs managed by Nomad is persisted to a server’s data directory.

Reviewer: ```suggestion
    jobs managed by Nomad is persisted to a server's data directory.
```

* **Access (read or write) to the Nomad configuration directory** - Access to

>     Nomad’s configuration file(s) directory can enable and disable features for

Reviewer: ```suggestion
    Nomad's configuration file(s) directory can enable and disable features for
```

Also worth noting that on server agents this can expose highly-privileged Vault and Consul tokens.

    a cluster.

* **Memory access to a running Nomad server agent** - Direct access to the

>     memory of the Nomad server agent process ( usually requiring a shell on the
>     system through various means ) results in almost all aspects of the agent

Reviewer: ```suggestion
    memory of the Nomad server agent process (usually requiring a shell on the
    system through various means) results in almost all aspects of the agent
```

    being compromised including access to certificates and other secrets.

The following are not part of the threat model for client agents:

* **Access (read or write) to the Nomad data directory** - Information about the
    allocations scheduled to a Nomad client is persisted to its data directory.

>     This would include any secrets in any of the allocation’s file systems.

Reviewer: ```suggestion
    This would include any secrets in any of the allocation's file systems.
```

* **Access (read or write) to the Nomad configuration directory** - Access to a

>     client’s configuration file can enable and disable features for a client

Reviewer: ```suggestion
    client's configuration file can enable and disable features for a client
```

    including insecure drivers such as
    [raw_exec](/docs/drivers/raw_exec.html).

* **Memory access to a running Nomad client agent** - Direct access to the
    memory of the Nomad client agent process allows an attack to extract secrets
    from clients such as Vault tokens.

* **Lax Client Driver Sandbox** - Drivers may allow some privileged operations,
    e.g. filesystem access to configuration directories, or raw accesses to host
    devices. Such privileges can be used to facilitate compromise other workloads,
    or cause denial-of-service attacks.

#### Internal Threats

> * **Operator** - Someone with a valid mTLS cert and ACL token may still be a

Reviewer: ```suggestion
* **Job Operator** - Someone with a valid mTLS certificate and ACL token may still be a
```

    threat to your cluster in certain situations, especially in multi-team
    cluster deployments. They may accidentally or intentionally use a malicious

>     jobspec to harm a cluster which can help be protected against using

Reviewer: ```suggestion
    job to harm a cluster which can help be protected against using
```

>     Namespaces and Sentinel policies.

Reviewer: Quotas and Sentinel are probably the biggest thing to mention here as without Quotas any job operator can exhaust all cluster resources.

* **Workload** - Workloads may have host network access within a cluster which
    can lead to SSRF due to application security issues outside of the scope of
    Nomad which may lead to internal access within the cluster. Using mTLS, ACLs

>     and Sentinel policies together can add layers of protection against

Reviewer: * mTLS should protect Nomad agents from SSRF
* Connect (which has mTLS) should protect Nomad-managed _services_ from SSRF

I don't think ACLs or Sentinel factor into SSRF.

We may want to break this down into `Service-to-agent communication` and `Service-to-service communication` as mTLS prevents the former, while Connect/Sentinel protect the latter.

Author: I see ACLs valuable here.  However, we might be splitting hairs here and arguing relative benefits of each, which is probably not very fruitful discussion.  I'm OK framing it as you wish.

IIUC, the concern with SSRF is an attacker causing a workload (e.g. golang proxy service) to issue an API request to the Nomad agent or another service, which the attacker doesn't have direct network access to.  Without mTLS, ACLs would reject these unauthenticated requests.  mTLS will help here indeed, assuming that the proxy doesn't initiate connections with the client certificate.

I don't know how Connect mTLS helps Nomad managed services; if Consul connect automatically setups mTLS sessions between services A<->B; then service B will vulnerable to [confused deputy problem](https://en.wikipedia.org/wiki/Confused_deputy_problem), and Consul Connect/mTLS wouldn't help if an attacker forges a request from A to B, as the request connection will be automatically wrapped with mTLS.  Here, service B must have another layer of request ACL/capability checking.

    malicious workloads.

* **RPC / API Access** - RPC and HTTP API endpoints without mTLS can expose
    clusters to abuse within the cluster from malicious workloads.

* **Client driver** - Drivers implement various workload types for a cluster,
    and the backend configuration of these drivers should be considered to
    implement defense in depth. For example, a custom Docker driver that limits
    the ability to mount the host file system may be subverted by network access

>     to an exposed Docker daemon API through other means such as the raw_exec

Reviewer: ```suggestion
    to an exposed Docker daemon API through other means such as the `raw_exec`
```

    driver.

#### External Threats

There are two main components to consider to for external threats in a Nomad cluster:

* **Server agent** - Internal cluster leader elections and replication is
    managed via Raft between server agents encrypted in transit. However,
    information about the server is stored unencrypted at rest in the agent’s
    data directory. This information may contain information such as ACL tokens
    and TLS certificates.

* **Client agent** - Client-to-server communication within a cluster is
    encrypted and authenticated using mTLS. Information about the allocations on
    a client node is unencrypted in the agent’s data and configuration
    directory.

### Network Ports

> 

Reviewer: ```suggestion
```

| **Port / Protocol**  | Agents  | Description |
|----------------------|---------|-------------|

> | **4646** / TCP       | All     | [HTTP](https://en.wikipedia.org/wiki/Hypertext_Transfer_Protocol) to provide [UI](/guides/web-ui/access.html) and [API](/api/index.html) access to agents. |

Reviewer: I think this is the canonical path? I would imagine we have to maintain the redirects ~forever so it probably doesn't matter.

```suggestion
| **4646** / TCP       | All     | [HTTP](https://en.wikipedia.org/wiki/Hypertext_Transfer_Protocol) to provide [UI](/guides/web-ui/access.html) and [API](/api-docs) access to agents. |
```
